rfcs/0075-declarative-wrappers.md

```diff
@@ -0,0 +1,353 @@
+ 
+```
+$ nix why-depends -f. beets gobject-introspection.dev
+/nix/store/93lfrhm8vp17m8ziqi8vp6v4cff67wkb-beets-1.4.9
+╚═══bin/beet: …-expat-2.2.8-dev/bin:/nix/store/y3ym76wrak3300vsjyf3klr52cnzmxwd-gobject-introspection-1.64.1-de…
+    => /nix/store/y3ym76wrak3300vsjyf3klr52cnzmxwd-gobject-introspection-1.64.1-dev
+```
+
+## Other issues
+
+- [issue 60260](https://github.com/NixOS/nixpkgs/issues/60260)
+
+General, justified complaint about wrappers.
+
+- [issue 95027](https://github.com/NixOS/nixpkgs/issues/95027)
+- [issue 23018](https://github.com/NixOS/nixpkgs/issues/23018)
+- [issue 11133](https://github.com/NixOS/nixpkgs/issues/11133)
+- [pull 95569](https://github.com/NixOS/nixpkgs/pull/95569)
+
+Since our wrappers are shell scripts, `gdb` can't run them. What if we had
+written a C based wrapper, that perhaps would read what environment it needs to
+set from a JSON file, and it will call the unwrapped original executable? I
+need feedback regarding whether `gdb` will play nice with this.
+
+This issue may not directly relate to declarative wrappers, and it is already
+addressed in @FRidh's [pull 95569](https://github.com/NixOS/nixpkgs/pull/95569), but perhaps
+both ideas could be integrated into an alternative, simpler creation method of
+binary wrappers. See [my
+comment](https://github.com/NixOS/nixpkgs/pull/95569#issuecomment-674508806).
+
+# Detailed design
+[design]: #detailed-design
+
+The current design is roughly implemented at
+[pull 85103](https://github.com/NixOS/nixpkgs/pull/85103) .  
+
+The idea is to have a Nix function, let us call it `wrapGeneric`, with an
+interface similar to
+[`wrapMpv`](https://github.com/NixOS/nixpkgs/blob/a5985162e31587ae04ddc65c4e06146c2aff104c/pkgs/applications/video/mpv/wrapper.nix#L9-L23)
+and
+[`wrapNeovim`](https://github.com/NixOS/nixpkgs/blob/a5985162e31587ae04ddc65c4e06146c2aff104c/pkgs/applications/editors/neovim/wrapper.nix#L11-L24)
+which will accept a single derivation or an array of them and will wrap all of
+their executables with the proper environment, based on their inputs.
+
+`wrapGeneric` should iterate recursively all `buildInputs` and
+`propagatedBuildInputs` of the input derivation(s), and construct an attrset with
+which it'll calculate the necessary environment of the executables. Then either
+via `wrapProgram` or a better method, it'll create the wrappers.
+
+A contributor using `wrapGeneric` shouldn't _care_ what type of wrapping needs
+to be performed on his derivation's executables - whether these are Qt related
+wrappings or a Gtk / gobject related. `wrapGeneric` should know all there is to
+know about environment variables every library / input may need during runtime,
+and with this information at hand, construct the necessary wrapper.
+
+In order for `wrapGenric` to know all of this information about our packaged
+libraries - the information about runtime env, we need to write in the
+`passthru`s of these libraries, what env vars they need. Such information was
```

Comment:
Limiting ourselves to environment variables is at times too limiting. Also, environment variables should be the implementation detail, although the variable names could be used as identifier.

##### Python example.
If you have a closure of packages, and there happens to be a Python application with Python libraries as dependencies, then that group of packages needs to be composed. If there is another Python application depending on other libraries, then that needs to be composed separately. At least, they are not allowed to share the same `*PYTHONPATH`. (Note one also should not use `PYTHONPATH` but that's a different discussion.)

Reply:
> If there is another Python application depending on other libraries, then that needs to be composed separately. At least, they are not allowed to share the same *PYTHONPATH

I see this is all about ABI compatibility, and I understand the issue while I remember you mentioned this in https://github.com/NixOS/nixpkgs/pull/85103 . I'm still unsure whether it's possible it'd happen, assuming our libraries are composed OK. @FRidh could you perhaps give a situation (with perhaps some rough code examples) to a situation where say Python2's and Python3's libraries will share the same `*PYTHONPATH`? It's important to note that the current design counts upon `buildInputs` and `propagatedBuildInputs` only.

Reply:
Different Python version is one case, the other is simply applications requiring different versions of a dependency.

Reply:
This can happen for example when you have a python3 application that shells out to an python2 application.

Reply:
Currently we prevent some of these effects by unsetting `NIX_PYTHONPATH` when a Python program, thereby avoid leakage. This problem with wrappers and environment variables should also be noted clearly in the RFC.

Reply:
OK I think I start to understand better the concerns. But still, this is not a "Drawback" of the idea, I mean, we still need to workaround such edge cases specifically in our current methods right? I agree though that it might be more complicated if someone is obliged to use `wrapGeneric` and not a wrap hook.

As an implementation detail, it may be worth noting that it'd be nice to make `wrapGeneric` able to "catch" such incompatible values in an environment variable.